pkg/chaosctl/common/exec.go

```diff
@@ -0,0 +1,104 @@
+	"google.golang.org/grpc/grpclog"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/remotecommand"
+	kubectlscheme "k8s.io/kubectl/pkg/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+// Exec executes certain command and returns the result
+// Only commands in chaos-mesh components should use this way
+// for target pod, use ExecBypass
+func Exec(ctx context.Context, pod v1.Pod, cmd string, c *kubernetes.Clientset) (string, error) {
+	name := pod.GetObjectMeta().GetName()
+	namespace := pod.GetObjectMeta().GetNamespace()
+	// TODO: if `containerNames` is set and specific container is injected chaos,
+	// need to use THE name rather than the first one.
+	// till 20/11/10 only podchaos and kernelchaos support `containerNames`, so not set it for now
+	containerName := pod.Spec.Containers[0].Name
+
+	req := c.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(name).
+		Namespace(namespace).
+		SubResource("exec")
+
+	req.VersionedParams(&v1.PodExecOptions{
+		Container: containerName,
+		Command:   []string{"/bin/sh", "-c", cmd},
+		Stdin:     false,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       false,
+	}, kubectlscheme.ParameterCodec)
+
+	var stdout, stderr bytes.Buffer
+	exec, err := remotecommand.NewSPDYExecutor(config.GetConfigOrDie(), "POST", req.URL())
+	if err != nil {
+		return "", fmt.Errorf("error in creating NewSPDYExecutor: %s", err.Error())
+	}
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+	if err != nil {
+		return "", fmt.Errorf("error in streaming remotecommand: %s, pod: %s, command: %s", err.Error(), pod.Name, cmd)
+	}
+	if stderr.String() != "" {
+		return "", fmt.Errorf("error of command %s: %s", cmd, stderr.String())
+	}
+	return stdout.String(), nil
+}
+
+// ExecBypass use chaos-daemon to enter namespace and execute command in target pod
+func ExecBypass(ctx context.Context, pod v1.Pod, daemon v1.Pod, cmd string, c *kubernetes.Clientset) (string, error) {
+	cmdSubSlice := strings.Fields(cmd)
+	if len(cmdSubSlice) == 0 {
+		return "", fmt.Errorf("command should not be empty")
+	}
+	// To disable printing irrelevant log from grpc/clientconn.go
+	// See grpc/grpc-go#3918 for detail. Could be resolved in the future
+	grpclog.SetLoggerV2(grpclog.NewLoggerV2(ioutil.Discard, ioutil.Discard, ioutil.Discard))
+	pid, err := GetPidFromPod(ctx, pod, daemon)
+	if err != nil {
+		return "", err
+	}
+	switch cmdSubSlice[0] {
+	case "ps":
+		nsenterPath := "-t " + strconv.Itoa(pid) + " -p -m"
+		newCmd := fmt.Sprintf("/usr/bin/nsenter %s -- /bin/sh -c '%s'", nsenterPath, cmd)
+		return Exec(ctx, daemon, newCmd, c)
+	case "cat", "ls":
+		// we need to enter mount namespace to get file related infomation
+		// but enter mnt ns would prevent us to access `cat`/`ls` in daemon
+		// so use `nsexec` to achieve using nsenter and cat together
+		newCmd := fmt.Sprintf("/usr/local/bin/nsexec %s %s", strconv.Itoa(pid), cmd)
```

Comment:
I came up with a better idea to implement a more general `nsexec` 🍻 . You can find the source code and documents at https://github.com/YangKeao/nsexec, and it will be brought to `chaos-daemon` in https://github.com/chaos-mesh/chaos-mesh/pull/1260/ .

I don't know whether it's a big project to migrate to [YangKeao/nsexec](https://github.com/YangKeao/nsexec).

Reply:
Cool I test it and it works fine. I'll commit the change till #1260 got merged.

Reply:
@YangKeao Updated to #1260, PTAL